test/affinity.cxx

```diff
@@ -0,0 +1,430 @@
+	REQUIRE(thread_policy_get(machThread, THREAD_AFFINITY_POLICY, reinterpret_cast<thread_policy_t>(&policy), &numPolicies, &getDefault) == 0);
+	return static_cast<uint32_t>(policy.affinity_tag);
+};
+#endif
+
+TEST_CASE("construct", "[affinity_t]")
+{
+	const auto affinity{substrate::make_unique_nothrow<substrate::affinity_t>()};
+	REQUIRE(affinity);
+}
+
+TEST_CASE("processor count", "[affinity_t]")
+{
+	const auto affinity{substrate::make_unique_nothrow<substrate::affinity_t>()};
+	REQUIRE(affinity);
+	REQUIRE(affinity->numProcessors() == std::thread::hardware_concurrency());
+}
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+TEST_CASE("iteration", "[affinity_t]")
+{
+	const auto affinity{substrate::make_unique_nothrow<substrate::affinity_t>()};
+	REQUIRE(affinity);
+#if defined(_POSIX_THREADS) && defined(_GNU_SOURCE)
+	cpu_set_t affinitySet{};
+	REQUIRE(sched_getaffinity(0, sizeof(cpu_set_t), &affinitySet) == 0);
+
+	REQUIRE(affinity->begin() != affinity->end());
+	std::size_t count{0};
+	for (const auto processor : *affinity)
+	{
+		++count;
+		REQUIRE(CPU_ISSET(processor, &affinitySet));
+	}
+	REQUIRE(count == affinity->numProcessors());
+#elif defined(_POSIX_THREADS) && defined(__APPLE__)
+	boolean_t getDefault{false};
+	thread_affinity_policy_data_t policy{};
+	mach_msg_type_number_t numPolicies{1};
+	const thread_port_t machThread{mach_thread_self()};
+	// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+	REQUIRE(thread_policy_get(machThread, THREAD_AFFINITY_POLICY, reinterpret_cast<thread_policy_t>(&policy), &numPolicies, &getDefault) == 0);
+	REQUIRE(affinity->begin() != affinity->end());
+	std::size_t count{0};
+	for (const auto processor: *affinity)
+	{
+		(void)processor;
+		++count;
+		REQUIRE(((policy.affinity_tag & 1 << count) || !policy.affinity_tag));
+	}
+	REQUIRE(count == affinity->numProcessors());
+#elif defined(_WIN32)
+	const auto procCount{GetActiveProcessorCount(ALL_PROCESSOR_GROUPS)};
+	REQUIRE(procCount > 0);
+	REQUIRE(affinity->numProcessors() == procCount);
+#endif
+}
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+TEST_CASE("pinning", "[affinity_t]")
+{
+	const auto affinity{substrate::make_unique_nothrow<substrate::affinity_t>()};
+	REQUIRE(affinity);
+#if defined(_POSIX_THREADS) && defined(_GNU_SOURCE)
+	for (const auto &processor : *affinity)
+	{
+		const auto result = std::async(std::launch::async, [&](const uint32_t currentProcessor) noexcept -> int32_t
+		{
+			try
+				{ affinity->pinThreadTo(currentProcessor); }
+			catch (const std::out_of_range &)
+				{ return -1; }
+			return sched_getcpu();
+		}, processor).get();
+		REQUIRE(result != -1);
+		REQUIRE(result == static_cast<int32_t>(processor));
+	}
+
+	REQUIRE(std::async(std::launch::async, [&]() noexcept -> bool
+	{
+		try { affinity->pinThreadTo(affinity->numProcessors()); }
+		catch (const std::out_of_range &) { return true; }
+		return false;
+	}).get());
```

Comment:
If it's possible to de-dupe this between the `#if` blocks here, that'd be great but no sweat if it can't be